File: 4设计四/svm.py
import numpy as np
import matplotlib.pyplot as plt
from sklearn import svm
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.datasets import make_blobs
from matplotlib import font_manager

# 设置中文字体
font_path = "C:/Windows/Fonts/msyh.ttc"
font_prop = font_manager.FontProperties(fname=font_path)
plt.rcParams['font.family'] = font_prop.get_name()
# 1. 生成可分线性数据（二维示例）
np.random.seed(1)  # 设置随机种子保证可重复性

# 生成两类数据（修正了center_box参数）
X, y = make_blobs(n_samples=100, centers=2, cluster_std=0.8,
                 center_box=(1.5, 3.5), random_state=1)
y = np.where(y == 0, -1, 1)  # 将标签转换为-1和1

# 可视化原始数据
plt.figure(figsize=(10, 6))
plt.scatter(X[y == 1, 0], X[y == 1, 1], c='b', marker='o', label='类别1')
plt.scatter(X[y == -1, 0], X[y == -1, 1], c='r', marker='+', label='类别2')
plt.title('原始数据分布')
plt.xlabel('特征1')
plt.ylabel('特征2')
plt.legend()
plt.grid(True)
plt.show()

# 2. 训练线性SVM模型
clf = svm.SVC(kernel='linear', C=1.0)
clf.fit(X, y)

# 获取模型参数
w = clf.coef_[0]  # 权重向量
b = clf.intercept_[0]  # 偏置项
sv = clf.support_vectors_  # 支持向量

print('SVM模型参数:')
print(f'权重向量 w = [{w[0]:.4f}, {w[1]:.4f}]')
print(f'偏置项 b = {b:.4f}')
print(f'支持向量数量: {len(sv)}')

# 3. 可视化决策边界和支持向量
plt.figure(figsize=(10, 6))

# 绘制数据点
plt.scatter(X[y == 1, 0], X[y == 1, 1], c='b', marker='o', label='类别1')
plt.scatter(X[y == -1, 0], X[y == -1, 1], c='r', marker='+', label='类别2')

# 绘制支持向量
plt.scatter(sv[:, 0], sv[:, 1], s=100, facecolors='none', edgecolors='k',
            linewidths=1.5, label='支持向量')

# 绘制决策边界
ax = plt.gca()
xlim = ax.get_xlim()
ylim = ax.get_ylim()

# 创建网格评估模型
xx = np.linspace(xlim[0], xlim[1], 100)
yy = np.linspace(ylim[0], ylim[1], 100)
YY, XX = np.meshgrid(yy, xx)
xy = np.vstack([XX.ravel(), YY.ravel()]).T
Z = clf.decision_function(xy).reshape(XX.shape)

# 绘制决策边界和间隔
ax.contour(XX, YY, Z, colors='k', levels=[-1, 0, 1], alpha=0.5,
           linestyles=['--', '-', '--'])
ax.scatter(clf.support_vectors_[:, 0], clf.support_vectors_[:, 1], s=100,
           linewidth=1, facecolors='none', edgecolors='k')

plt.title('线性SVM分类结果')
plt.xlabel('特征1')
plt.ylabel('特征2')
plt.legend()
plt.grid(True)
plt.show()

# 4. 模型评估
y_pred = clf.predict(X)
accuracy = accuracy_score(y, y_pred)
print(f'\n模型准确率: {accuracy:.2%}')

print('混淆矩阵:')
print(confusion_matrix(y, y_pred))

# 5. 新样本预测示例
new_X = np.array([[2.5, 3.0],
                 [1.8, 2.2],
                 [3.2, 3.5]])
new_y_pred = clf.predict(new_X)

print('\n新样本预测结果:')
for i, (sample, pred) in enumerate(zip(new_X, new_y_pred)):
    print(f'样本 {sample} -> 预测类别: {pred}')

File: 4设计四/svm.m
%% 线性SVM分类案例 - MATLAB实现
clc; clear; close all;

%% 1. 生成可分线性数据（二维示例）
rng(1); % 设置随机种子保证可重复性

% 类别1数据
n1 = 50;
X1 = 1.5 + 0.8*randn(n1,2);

% 类别2数据
n2 = 50;
X2 = 3.5 + 0.8*randn(n2,2);

% 合并数据
X = [X1; X2];
y = [ones(n1,1); -ones(n2,1)];

% 可视化原始数据
figure;
gscatter(X(:,1), X(:,2), y, 'rb', 'o+');
title('原始数据分布');
xlabel('特征1'); ylabel('特征2');
legend('类别1', '类别2');
grid on;

%% 2. 训练线性SVM模型
svmModel = fitcsvm(X, y, 'KernelFunction', 'linear', ...
    'BoxConstraint', 1, 'Standardize', true);

% 获取模型参数
w = svmModel.Beta;  % 权重向量
b = svmModel.Bias;  % 偏置项
sv = svmModel.SupportVectors; % 支持向量

fprintf('SVM模型参数:\n');
fprintf('权重向量 w = [%.4f, %.4f]\n', w(1), w(2));
fprintf('偏置项 b = %.4f\n', b);
fprintf('支持向量数量: %d\n', size(sv,1));

%% 3. 可视化决策边界和支持向量
figure;
% 绘制数据点
gscatter(X(:,1), X(:,2), y, 'rb', 'o+');
hold on;

% 绘制支持向量
plot(sv(:,1), sv(:,2), 'ko', 'MarkerSize', 10, 'LineWidth', 1.5);

% 绘制决策边界
x1 = linspace(min(X(:,1))-0.5, max(X(:,1))+0.5, 100);
x2 = -(w(1)*x1 + b)/w(2);
plot(x1, x2, 'k-', 'LineWidth', 2);

% 绘制间隔边界
plot(x1, -(w(1)*x1 + b + 1)/w(2), 'k--');
plot(x1, -(w(1)*x1 + b - 1)/w(2), 'k--');

title('线性SVM分类结果');
xlabel('特征1'); ylabel('特征2');
legend('类别1', '类别2', '支持向量', '决策边界', '间隔边界');
grid on;
hold off;

%% 4. 模型评估
% 预测训练数据
yPred = predict(svmModel, X);

% 计算准确率
accuracy = sum(yPred == y)/length(y);
fprintf('\n模型准确率: %.2f%%\n', accuracy*100);

% 混淆矩阵
confMat = confusionmat(y, yPred);
disp('混淆矩阵:');
disp(confMat);

%% 5. 新样本预测示例
newX = [2.5, 3.0;
        1.8, 2.2;
        3.2, 3.5];
newYPred = predict(svmModel, newX);

fprintf('\n新样本预测结果:\n');
for i = 1:size(newX,1)
    fprintf('样本 [%.2f, %.2f] -> 预测类别: %d\n', ...
            newX(i,1), newX(i,2), newYPred(i));
end